wsrep_config_service.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2022 Codership Oy <info@codership.com>
+ *
+ * This file is part of wsrep-API.
+ *
+ * Wsrep-API is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Wsrep-API is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with wsrep-API.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/** @file wsrep_config_service.h
+ *
+ * This file defines interface to retrieve a complete list of configuration
+ * parameters accepted by the provider.
+ * *
+ * The provider which is capable of using the service interface v1 must
+ * export the following functions:
+ *
+ * int wsrep_init_config_service_v1(wsrep_config_service_v1_t*)
+ * void wsrep_deinit_config_service_v1()
+ *
+ * which can be probed by the application.
+ *
+ */
+
+#ifndef WSREP_CONFIG_SERVICE_H
+#define WSREP_CONFIG_SERVICE_H
+
+#include "wsrep_api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Flags to describe parameters.
+ * By default, a parameter is dynamic and of type string,
+ * unless flagged otherwise.
+ */
+#define WSREP_PARAM_DEPRECATED    (1 << 0)
+#define WSREP_PARAM_READONLY      (1 << 1)
+#define WSREP_PARAM_TYPE_BOOL     (1 << 2)
+#define WSREP_PARAM_TYPE_INTEGER  (1 << 3)
+#define WSREP_PARAM_TYPE_DOUBLE   (1 << 4)
+
+#define WSREP_PARAM_TYPE_MASK ( \
+    WSREP_PARAM_TYPE_BOOL     | \
+    WSREP_PARAM_TYPE_INTEGER  | \
+    WSREP_PARAM_TYPE_DOUBLE     \
+    )
+
+typedef struct wsrep_parameter
+{
+  int flags;
+  const char* name;
+  union {
+    bool as_bool;
+    int64_t as_integer;
+    double as_double;
+    const char* as_string;
+  } value;
+} wsrep_parameter_t;
+
+/**
+ * Callback called once for each parameter exposed by provider.
+ *
+ * @param p        parameter
+ * @param context  application context
+ */
+typedef void (*wsrep_get_parameters_cb) (const wsrep_parameter_t* p,
+                                         void* context);
+
+/**
+ * Get configuration parameters exposed by the provider.
+ *
+ * @param wsrep    pointer to provider handle
+ * @param cb       function pointer for callback
+ * @param context  application context passed to callback
+ */
+typedef void (*wsrep_get_parameters_fn) (wsrep_t* wsrep,
```

Comment:
This function returns void. How errors are propagated?

Reply:
We probably need to handle both error in provider and error on server side during callback call.

Reply:
Some error handling is probably missing... will fix.